### PR TITLE
Link teuchos and other libraries to csympy and make teuchos static

### PR DESCRIPTION
**Pull request by isuruf**
With this we can use `-DBUILD_SHARED_LIBS=on/off` to make csympy shared lib on/off.

### CODE REVIEW

**Comment by certik** [comment]
I just tested it and it works. Thanks! With `-DBUILD_SHARED_LIBS=off` it builds `libcsympy.a` and `libteuchos.a`. With `-DBUILD_SHARED_LIBS=on` it builds `libcsympy.so` and `libteuchos.a`.

One thing that is still not clear to me is if the `libteuchos.a` is always included in `libcsympy.a` now or not. In any case, I think we should now remove explicit linking with `teuchos`: https://github.com/isuruf/csympy/pull/3

**Comment by isuruf** [comment]
`libteuchos.a` is linked with `libcsympy.a`, but the problem when another program loads the csympy static library it won't load `teuchos` automatically. So, we have to give the `-lteuchos` argument.

On the other hand when we build the shared library `libcsympy.so` it will load `libteuchos.a` and all the other libraries it is linked with. This way, when we load the csympy library in julia, all other libraries are loaded.

**Comment by certik** [comment]
If we built a static library, then I think we need to link the `${LIBS}`, don't we?

**Comment by isuruf** [comment]
Yes. you are right. Will remove that commit.

**Comment by isuruf** [comment]
`target_link_libraries` is transitive. Since `teuchos` and `${LIBS}` are linked to csympy, tests and benchmarks are linked to  `teuchos` and `${LIBS}` as well.

**Comment by isuruf** [comment]
I added some tests for shared libs in travis. Any ideas on what the options should be when testing for shared libs?

File: bin/test_travis.sh [hunk 1]
```diff
@@ -98,8 +101,8 @@ else
     if [[ "${WITH_ARB}" != "" ]]; then
         extra_libs="$extra_libs -larb -lflint"
     fi
-    ${CXX} -std=c++0x -I$our_install_dir/include/ -L$our_install_dir/lib test_basic.cpp -lcsympy -lgmpxx -lgmp -lteuchos $extra_libs
-    export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH
+    ${CXX} -std=c++0x -I$our_install_dir/include/ -L$our_install_dir/lib test_basic.cpp -lcsympy -lteuchos -lgmpxx -lgmp $extra_libs
```
Comment (certik) [inline bin/test_travis.sh]: It's not a big deal, but why don't you move the `-lteuchos` to where it was (behind `-lgmp`), to reduce clutter in the diff?

Comment (isuruf) [inline bin/test_travis.sh]: Sure

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -47,6 +47,7 @@ configure_file(csympy_config.h.in csympy_config.h)
 include_directories(BEFORE ${csympy_BINARY_DIR}/src)
 
 add_library(csympy ${SRC})
+target_link_libraries(csympy teuchos ${LIBS})
```
Comment (certik) [inline src/CMakeLists.txt]: @isuruf is my understanding correct, that this line causes the following:

a) if `csympy` is a static lib, then cmake will add `teuchos` and `${LIBS}` into gcc options for each linking of `csympy`
b) if `csympy` is a dynamic lib, then `teuchos` (which is a static lib) will get included in it, as well as any static libs from `${LIBS}` and the rest (i.e. the dynamic libs) will be added to gcc link options.

Comment (isuruf) [inline src/CMakeLists.txt]: a) if `csympy` is a static lib, I think this changes nothing.
b) what you said is correct

**Comment by certik** [comment]
I think this PR looks good as it is. You can merge it. You don't need to move the `teuchos` thing, it makes sense to put it behind `csympy`, as it is.

File: src/CMakeLists.txt (hunk 2 above)
Comment (certik) [inline src/CMakeLists.txt]: For a), I think if you didn't put `teuchos` into `target_link_libraries`, then you need to explicitly add `teuchos` for each linking, as we did in tests for example. So I think it changes what libraries cmake appends during the link stage of programs that use csympy. I agree that it changes nothing how the `csympy.a` itself is linked.

Comment (isuruf) [inline src/CMakeLists.txt]: Yes, that's true

**Comment by isuruf** [comment]
Thanks for the review

**Comment by certik** [comment]
Build with `-DBUILD_SHARED_LIBS=on`, to get a shared library.

**Pull request merged**